src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs

```diff
@@ -38,5 +38,41 @@ public sealed override Task<int> ReadAsync(byte[] buffer, int offset, int count,
 
         public sealed override void CopyTo(Stream destination, int bufferSize) =>
             CopyToAsync(destination, bufferSize, CancellationToken.None).GetAwaiter().GetResult();
+
+        public virtual Task DrainAsync(int maxDrainBytes) => Task.CompletedTask;
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (_connection != null)
+                {
+                    // Start the asynchronous drain.
+                    // It may complete synchronously, in which case the connection will be put back in the pool synchronously.
+                    // Skip the call to base.Dispose -- it will be deferred until DrainOnDispose finishes.
+                    DrainOnDispose();
+                    return;
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        // Maximum request drain size, 16K.
+        private const int s_maxDrainBytes = 16 * 1024;
+
+        private async void DrainOnDispose()
+        {
+            try
+            {
+                await DrainAsync(s_maxDrainBytes);
+            }
+            catch (Exception)
+            {
+                // Eat any exceptions and just Dispose.
```

Comment:
It'd be good to log this.

Reply:
Sounds good

Reply:
Actually, I'm wondering exactly what to log here.

We may want to log when the drain is unsuccessful, i.e. we killed the connection because we exceeded the limit.  That seems most useful.

We could also log when we get an exception trying to do the drain (what you suggested above).  I'm not sure how useful this is, since it basically just indicates a server problem.  But there's no harm in logging it, so maybe we should just do it.

Reply:
I think we should log all of that:
- When we successfully drain, and potentially even how much we drained.
- When we failed to drain because there was too much and thus that's why we're disposing the connection.
- If the draining threw, what was the cause, i.e. the exception.